Some jobs cannot be sent to cluster due to ClassCastException

When job which uses some lambda is submitted to remote cluster then exception is thrown. 

What is interesting is that it happens only for some jobs with lambdas. For example in soak test suite most of the tests fail with that but for example `stateful-map-test` works correctly. However it can be simply reproduced by mentioned below reproducer.

This is regression caused by changes in https://github.com/hazelcast/hazelcast-jet/commit/2115dc70e1a518e405ddb05eeadb14910af6280c

Stacktrace:
```
com.hazelcast.jet.JetException: Class containing the lambda probably missing from class path, did you add it using JobConfig.addClass()?: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier.readerSupplier of type com.hazelcast.function.BiFunctionEx in instance of com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier
	at com.hazelcast.jet.impl.util.ExceptionUtil.handleSerializedLambdaCce(ExceptionUtil.java:218)
	at com.hazelcast.jet.impl.execution.init.CustomClassLoadedObject.deserializeWithCustomClassLoader(CustomClassLoadedObject.java:68)
	at com.hazelcast.jet.impl.JobCoordinationService.deserializeJobDefinition(JobCoordinationService.java:860)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitJob$1(JobCoordinationService.java:205)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$46(JobCoordinationService.java:1023)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$47(JobCoordinationService.java:1044)
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier.readerSupplier of type com.hazelcast.function.BiFunctionEx in instance of com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleException(SerializationUtil.java:103)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:209)
	at com.hazelcast.jet.impl.execution.init.CustomClassLoadedObject.deserializeWithCustomClassLoader(CustomClassLoadedObject.java:66)
	... 11 more
Caused by: java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier.readerSupplier of type com.hazelcast.function.BiFunctionEx in instance of com.hazelcast.jet.impl.connector.ReadMapOrCacheP$LocalProcessorMetaSupplier
	at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2287)
	at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1417)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2293)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at java.util.HashMap.readObject(HashMap.java:1407)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1170)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2178)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:83)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:76)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:203)
	... 12 more
```

**Steps to reproduce:**
It can be reproduced most of soak tests which can be run locally. We can reproduce it with `python-test` soak test since it is simple for configuration.
1. clone https://github.com/hazelcast/hazelcast-jet-ansible-tests and build it by `mvn clean install -DskipTests`
2. unzip latest version of Jet and copy `hazelcast-jet-python-4.3-SNAPSHOT.jar` from `opt` to `lib`
3. start Jet from `bin` directory by `./jet-start`
4. submit `python-test` from `bin` directory by `./jet submit -v ${PATH_TO_SOAK_TESTS_DIR}/python-test/target/python-test-4.3-SNAPSHOT.jar` where `${PATH_TO_SOAK_TESTS_DIR}` is path with your `hazelcast-jet-ansible-tests` from step 1

If you use Jet version before aforementioned commit then it works correctly.

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetBootstrap.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetBootstrap.java
@@ -33,6 +33,7 @@ import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.core.DAG;
 import com.hazelcast.jet.impl.util.ConcurrentMemoizingSupplier;
 import com.hazelcast.jet.impl.util.JetConsoleLogHandler;
+import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
 import com.hazelcast.map.IMap;
@@ -253,8 +254,18 @@ public final class JetBootstrap {
         }
 
         @Nonnull @Override
+        public Job newJob(@Nonnull Pipeline pipeline, @Nonnull JobConfig config) {
+            return instance.newJob(pipeline, updateJobConfig(config));
+        }
+
+        @Nonnull @Override
         public Job newJob(@Nonnull DAG dag, @Nonnull JobConfig config) {
             return instance.newJob(dag, updateJobConfig(config));
+        }
+
+        @Nonnull @Override
+        public Job newJobIfAbsent(@Nonnull Pipeline pipeline, @Nonnull JobConfig config) {
+            return instance.newJobIfAbsent(pipeline, updateJobConfig(config));
         }
 
         @Nonnull @Override

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/transform/GlobalFlatMapStatefulTransform.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/transform/GlobalFlatMapStatefulTransform.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.jet.impl.pipeline.transform;
 
+import com.hazelcast.function.SupplierEx;
 import com.hazelcast.function.ToLongFunctionEx;
 import com.hazelcast.jet.Traverser;
 import com.hazelcast.jet.function.TriFunction;
@@ -24,20 +25,19 @@ import com.hazelcast.jet.impl.pipeline.Planner.PlannerVertex;
 import com.hazelcast.jet.impl.util.ConstantFunctionEx;
 
 import javax.annotation.Nonnull;
-import java.util.function.Supplier;
 
 import static com.hazelcast.jet.core.processor.Processors.flatMapStatefulP;
 
 public class GlobalFlatMapStatefulTransform<T, S, R> extends AbstractTransform {
 
     private final ToLongFunctionEx<? super T> timestampFn;
-    private final Supplier<? extends S> createFn;
+    private final SupplierEx<? extends S> createFn;
     private final TriFunction<? super S, Object, ? super T, ? extends Traverser<R>> statefulFlatMapFn;
 
     public GlobalFlatMapStatefulTransform(
             @Nonnull Transform upstream,
             @Nonnull ToLongFunctionEx<? super T> timestampFn,
-            @Nonnull Supplier<? extends S> createFn,
+            @Nonnull SupplierEx<? extends S> createFn,
             @Nonnull TriFunction<? super S, Object, ? super T, ? extends Traverser<R>> statefulFlatMapFn
     ) {
         super("flatmap-stateful-global", upstream);

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/transform/GlobalMapStatefulTransform.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/transform/GlobalMapStatefulTransform.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.jet.impl.pipeline.transform;
 
+import com.hazelcast.function.SupplierEx;
 import com.hazelcast.function.ToLongFunctionEx;
 import com.hazelcast.jet.function.TriFunction;
 import com.hazelcast.jet.impl.pipeline.Planner;
@@ -23,20 +24,19 @@ import com.hazelcast.jet.impl.pipeline.Planner.PlannerVertex;
 import com.hazelcast.jet.impl.util.ConstantFunctionEx;
 
 import javax.annotation.Nonnull;
-import java.util.function.Supplier;
 
 import static com.hazelcast.jet.core.processor.Processors.mapStatefulP;
 
 public class GlobalMapStatefulTransform<T, S, R> extends AbstractTransform {
 
     private final ToLongFunctionEx<? super T> timestampFn;
-    private final Supplier<? extends S> createFn;
+    private final SupplierEx<? extends S> createFn;
     private final TriFunction<? super S, Object, ? super T, ? extends R> statefulMapFn;
 
     public GlobalMapStatefulTransform(
             @Nonnull Transform upstream,
             @Nonnull ToLongFunctionEx<? super T> timestampFn,
-            @Nonnull Supplier<? extends S> createFn,
+            @Nonnull SupplierEx<? extends S> createFn,
             @Nonnull TriFunction<? super S, Object, ? super T, ? extends R> statefulMapFn
     ) {
         super("map-stateful-global", upstream);

